pennylane/templates/layers.py

```diff
@@ -0,0 +1,418 @@
+    """A layer applying rotations on each qubit followed by cascades of 2-qubit entangling gates.
+
+    The 2-qubit or imprimitive gates act on each qubit :math:`i` chronologically. The second qubit for
+    each gate is determined by :math:`(i+r)\mod n`, where :math:`n` is the number of qubits
+    and :math:`r` a layer hyperparameter called the range.
+
+    This is an example of two 4-qubit strongly entangling layers (ranges :math:`r=1` and :math:`r=2`) with
+    rotations :math:`R` and CNOTs as imprimitives:
+
+    .. figure:: ../../_static/layer_sec.png
+        :align: center
+        :width: 60%
+        :target: javascript:void(0);
+
+    Args:
+        weights (array[float]): array of weights of shape ``(len(wires), 3)``
+
+    Keyword Args:
+        r (Sequence[int]): range of the imprimitive gates of this layer
+        imprimitive (pennylane.ops.Operation): two-qubit gate to use, defaults to :class:`~.CNOT`
+        wires (Sequence[int]): sequence of qubit indices that the template acts on
+    """
+    if len(wires) < 2:
+        raise ValueError("StronglyEntanglingLayer requires at least two wires or subsystems to apply "
+                         "the imprimitive gates.")
+
+    for i, wire in enumerate(wires):
+        Rot(weights[i, 0], weights[i, 1], weights[i, 2], wires=wire)
+
+    num_wires = len(wires)
+    for i in range(num_wires):
+        imprimitive(wires=[wires[i], wires[(i+r) % num_wires]])
+
+
+def RandomLayers(weights, ratio_imprim=0.3, imprimitive=CNOT, rotations=[RX, RY], wires=None):
+    """A sequence of layers of type :func:`RandomLayer()`.
+
+    The number of layers :math:`L` and the number :math:`k` of rotations per layer is inferred from the first
+    and second dimension of ``weights``. The type of imprimitive (two-qubit) gate and rotations distributed
+    randomly in the circuit can be chosen explicitly.
+
+    Args:
+        weights (array[float]): array of weights of shape ``(L, k)``,
+
+    Keyword Args:
+        ratio_imprim (float): value between 0 and 1 that determines the ratio of imprimitive to rotation gates
+        imprimitive (pennylane.ops.Operation): two-qubit gate to use, defaults to :class:`~.CNOT`
+        rotations (list[str]): List of Pauli-X, Pauli-Y and/or Pauli-Z gates. The frequency determines how often
+            a particular rotation type is used.
+        wires (Sequence[int]): sequence of qubit indices that the template acts on
+    """
+
+    for layer_weights in weights:
+        RandomLayer(layer_weights, ratio_imprim=ratio_imprim, imprimitive=imprimitive, rotations=rotations, wires=wires)
+
+
+def RandomLayer(weights, ratio_imprim=0.3, imprimitive=CNOT, rotations=[RX, RY], wires=None):
+    """A layer of randomly chosen single qubit rotations and 2-qubit entangling gates, acting
+    on randomly chosen qubits.
+
+    The number :math:`k` of single qubit rotations is inferred of the first dimension of ``weights``.
```

Comment:
The number k of single qubit rotations is inferred **from** the first dimension of weights

Reply:
Done